[Enhancement]generate new range faster

zombee0 opened the pull request:
## Why I'm doing:
I checked assembly code, complier can't generate branch-less code, 
and it's not critical part in cpu usage, so just use simple way

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

zombee0 committed:
[Enhancement]generate new range faster

Signed-off-by: zombee0 <ewang2027@gmail.com>

alvin-celerdata commented on be/src/storage/range.h:
```diff
@@ -89,9 +89,10 @@ inline Range<T> Range<T>::filter(const Filter* const filter) const {
     int32_t start = filter->size();
     int32_t end = -1;
-    for (int32_t i = 0; i < filter->size(); i++) {
+    for (int32_t i = 0; i < filter->size() && start > end; i++) {
```
better assign `filter->size()` to a const local variable, which is good for the compiler to optimize

zombee0 replied:
will do it, but i think the compiler can also do the same optimization on this code.

trueeyu commented on be/src/storage/range.h:
```diff
@@ -89,9 +89,10 @@ inline Range<T> Range<T>::filter(const Filter* const filter) const {
         start = start > i && filter->data()[i] == 1 ? i : start;
-        end = end < i && filter->data()[i] == 1 ? i : end;
+        int32_t tail_index = filter->size() - 1 - i;
+        end = end < tail_index && filter->data()[tail_index] == 1 ? tail_index : end;
```
What is this function used for? I couldn't understand its meaning just by looking at the code. Is it for finding the upper and lower bounds? Why not just write two loops—one for finding the start and one for finding the end?

zombee0 replied:
Yes, it's used to find upper and lower bounds, i think less loop, less branch, we also use `?:` to reduce branch.

trueeyu replied:
? is also one branch?

zombee0 replied:
i think it can be optimized as cmov

zombee0 replied:
I checked assembly code
complier cann't generate branch-less code, and it's not critical part in cpu usage,
so just use simple way

zombee0 committed:
address comments

Signed-off-by: zombee0 <ewang2027@gmail.com>

zombee0 committed:
checked assembly code

complier cann't generate branch-less code,
and it's not critical part in cpu usage,
so just use simple way

Signed-off-by: zombee0 <ewang2027@gmail.com>
